server/channels/app/channel.go

```diff
@@ -3632,7 +3633,74 @@ func (a *App) ConvertGroupMessageToChannel(c request.CTX, convertedByUserId stri
 	_ = a.postMessageForConvertGroupMessageToChannel(c, gmConversionRequest.ChannelID, convertedByUserId, users)
 
 	// the user conversion the GM becomes the channel admin.
-	_, appErr = a.UpdateChannelMemberSchemeRoles(c, gmConversionRequest.ChannelID, convertedByUserId, false, true, true)
+	if convertedByUserId != "" {
+		_, appErr = a.UpdateChannelMemberSchemeRoles(c, gmConversionRequest.ChannelID, convertedByUserId, false, true, true)
+		if appErr != nil {
+			return nil, appErr
+		}
+	} else {
+		// if there is no one converting the GM, then the everyone should become the channel admin.
+		for _, user := range users {
+			_, appErr = a.UpdateChannelMemberSchemeRoles(c, gmConversionRequest.ChannelID, user.Id, false, true, true)
+			if appErr != nil {
+				return nil, appErr
+			}
+		}
+	}
+
+	return updatedChannel, nil
+}
+
+func (a *App) convertDirectMessageToChannel(c request.CTX, leavingUsername string, channelID string) (*model.Channel, *model.AppError) {
+	originalChannel, appErr := a.GetChannel(c, channelID)
+	if appErr != nil {
+		return nil, appErr
+	}
+
+	if originalChannel.Type != model.ChannelTypeDirect {
+		return nil, model.NewAppError("ConvertDirectMessageToChannel", "app.channel.direct_message_conversion.original_channel_not_dm", nil, "", http.StatusBadRequest)
+	}
+
+	if leavingUsername == "" {
+		leavingUsername = "Deleted User"
+	}
+
+	toUpdate := originalChannel.DeepCopy()
+	toUpdate.Type = model.ChannelTypePrivate
+	toUpdate.TeamId = originalChannel.TeamId
+	toUpdate.Name = originalChannel.Name
+	toUpdate.DisplayName = leavingUsername
+
+	updatedChannel, appErr := a.UpdateChannel(c, toUpdate)
+	if appErr != nil {
+		return nil, appErr
+	}
+
+	users, appErr := a.GetUsersInChannelPage(&model.UserGetOptions{
+		InChannelId: channelID,
+		Page:        0,
+		PerPage:     2,
+	}, false)
+	if appErr != nil {
+		return nil, appErr
+	}
+
+	a.Srv().Platform().InvalidateCacheForChannel(originalChannel)
+
+	_ = a.setSidebarCategoriesForConvertedGroupMessage(c, &model.GroupMessageConversionRequestBody{
+		TeamID: originalChannel.TeamId,
+	}, users)
+
+	var remainingUserId string
+	for _, user := range users {
+		if user.Username != leavingUsername {
+			remainingUserId = user.Id
+			break
+		}
+	}
```

Comment:
Do we really need a for loop for a 2 people channel?

Reply:
Not really, but it's easier to read for me :p Also panic safe.